File: dart/2901.dart
class Solution {
    bool validHam(String i, String j) {
        if (i.length != j.length)
            return false;
        int n = i.length;
        bool f = false;
        for (int k = 0; k < n; k++)
            if (i[k] != j[k]) {
                if (f)
                    return false;
                f = true;
            }
        return true;
    }

    void find(List<String> a, List<int> g, int i, int n, List<List<String>> l) {
        if (i == n)
            return;
        List<String> t = l[i];
        t.add(a[i]);
        find(a, g, i + 1, n, l);
        List<String> p = [];
        for (int k = i + 1; k < n; k++)
            if (g[i] != g[k] && validHam(a[i], a[k]) && p.length < l[k].length)
                p = l[k];
        for (String k in p)
            t.add(k);
    }

    List<String> getWordsInLongestSubsequence(List<String> w, List<int> g) {
        int n = w.length;
        List<List<String>> d = [];
        for (int k = 0; k < n; k++)
            d.add([]);
        find(w, g, 0, n, d);
        List<String> t = d[0];
        for (List<String> k in d)
            if (t.length < k.length)
                t = k;
        return t;
    }
}

File: cpp/2901.cpp
class Solution {
private:
    bool validHam(string& i, string& j) {
        if (i.length() != j.length())
            return false;
        int n = i.length();
        bool f = false;
        for (int k = 0; k < n; k++)
            if (i[k] != j[k]) {
                if (f)
                    return false;
                f = true;
            }
        return true;
    }

private:
    void find(vector<string>& a, vector<int>& g, int i, int n,
              vector<vector<string>>& l) {
        if (i == n)
            return;
        l[i].push_back(a[i]);
        find(a, g, i + 1, n, l);
        int p = -1;
        for (int k = i + 1; k < n; k++)
            if (g[i] != g[k] && validHam(a[i], a[k]) &&
                (p == -1 || l[p].size() < l[k].size()))
                p = k;
        if (p != -1)
            for (string& s : l[p])
                l[i].push_back(s);
    }

public:
    vector<string> getWordsInLongestSubsequence(vector<string>& w,
                                                vector<int>& g) {
        int n = w.size();
        vector<vector<string>> d(n, vector<string>());
        find(w, g, 0, n, d);
        int t = 0;
        for (int i = 1; i < n; i++)
            if (d[t].size() < d[i].size())
                t = i;
        return d[t];
    }
};
